### DISCUSSION TITLE
Sending s3 audit logs to td-agent

### DISCUSSION THREAD

**Discussion by aoberest** [seaweedfs/seaweedfs#2594] [Q&A] [answered]
Hello,
I'm trying to set up sending s3 audit logs through a locally installed td-agent
I have read https://github.com/chrislusf/seaweedfs/wiki/S3-API-Audit-log and want to make another scheme for sending s3 audit logs.

Seaweedfs filer --> local td-agent --> remote logstash --> Elasticsearch

/etc/seaweedfs/auditLogConfig.json

```
{
    "fluent_port": 24224,
    "fluent_host": "127.0.0.1",
    "fluent_network": "tcp",
    "timeout": 3,
    "write_timeout": 0,
    "buffer_limit": 8192,
    "retry_wait": 500,
    "max_retry": 13,
    "max_retry_wait": 60000,
    "tag_prefix": "myindex.service.seaweedfs.s3auditlog"
}
```

td-agent.conf

```
  <source>
    @type forward
    port 24224
    bind "127.0.0.1"

  </source>
  <filter myindex.service.seaweedfs.s3auditlog.**>
    @type parser
    key_name "requester"
    <parse>
      @type "json"
      time_key "time"
      time_type unixtime
      keep_time_key true
    </parse>
  </filter>
  <filter myindex.**>
    @type record_transformer
    enable_ruby 
    <record>
      original_time ${ record.key?("@timestamp") ? DateTime.parse(record["@timestamp"]).iso8601(3) : time.iso8601(3) }
    </record>
  </filter>
  <match myindex.**>
    @type forward
    time_as_integer true
    transport tcp
    <server>
      host "secret"
      port 24224
    </server>
    <buffer>
      @type "file"
      path "/var/cache/td-agent/buffer/"
      total_limit_size 4GB
      flush_interval 10s
      flush_at_shutdown true
      flush_thread_count 8
    </buffer>
  </match>
```

starting fluentd-1.0.2
listening port port=24224 bind="127.0.0.1"

After some time i got warning in filler log 

`W0117 13:09:27 18092 audit_fluent.go:73] Error while posting log: fluent#write: could not connect to fluentd after 13 retries`

After that i added @log_level trace in input source td-agent and did not see any request.... 

Сan anyone help me with this question?

@kmlebedev Maybe you will find time to look at this case?

**Comment by kmlebedev** [comment, marked as answer]
You can see a working [example](https://github.com/chrislusf/seaweedfs/blob/master/docker/compose/local-auditlog-compose.yml) and add your agent to it, if the problem reproduces please provide your docker compose file

**Reply by kmlebedev** [reply]
@aoberest Why don't you send logs directly to the logstash via the fluent plugin/codec ?
PS. in Russian can be discussed at https://seaweedfs.slack.com/archives/C9MGUC1UG

**Reply by aoberest** [reply]
Thanks for the answer. Tomorrow I will try docker compose file and and I'll let you know the result.

**Reply by aoberest** [reply]
@kmlebedev Hello

I seem to have found an error.
First I ran everything by default and it works.
Then I took fluent.json from the wiki and got an error.
If you do not specify `"timeout": 3` then the logs are sent, if you specify, then I get an error.

```
s3_1      | [2022-01-18T09:50:29Z] Unable to send logs to fluentd, reconnecting...
s3_1      | W0118 09:50:29     1 audit_fluent.go:73] Error while posting log: fluent#write: could not connect to fluentd after 13 retries
s3_1      | [2022-01-18T09:51:53Z] Unable to send logs to fluentd, reconnecting...
s3_1      | W0118 09:51:53     1 audit_fluent.go:73] Error while posting log: fluent#write: could not connect to fluentd after 13 retries
s3_1      | [2022-01-18T09:53:16Z] Unable to send logs to fluentd, reconnecting...
s3_1      | W0118 09:53:16     1 audit_fluent.go:73] Error while posting log: fluent#write: could not connect to fluentd after 13 retries
s3_1      | [2022-01-18T09:54:40Z] Unable to send logs to fluentd, reconnecting...
s3_1      | W0118 09:54:40     1 audit_fluent.go:73] Error while posting log: fluent#write: could not connect to fluentd after 13 retries
```

Replace https://github.com/chrislusf/seaweedfs/blob/master/docker/compose/fluent.json 
fluent.json
```
{
    "fluent_port": 24224,
    "fluent_host": "fluent",
    "fluent_network": "tcp",
    "timeout": 3,
    "write_timeout": 0,
    "buffer_limit": 8192,
    "retry_wait": 500,
    "max_retry": 13,
    "max_retry_wait": 60000,
    "tag_prefix": ""
}
```
and try to test.

P.S. I couldn't connect to slack. 
I have telegram SeaweedFS chat https://t.me/Seaweedfs
I will be glad if you join us

**Reply by kmlebedev** [reply]
try setting `"timeout": 3000,` 
in milliseconds

**Reply by aoberest** [reply]
Working option for me.

auditLogConfig.json
```
{
    "fluent_port": 24224,
    "fluent_host": "127.0.0.1",
    "tag_prefix": ""
}
```

addition td-agent options
/etc/td-agent/conf.d/td-seaweedfs-s3auditlog.conf

```
# Audit s3 fluentd format log
<source>
  @type forward
  port 24224
  bind 127.0.0.1
</source>
```

Os Centos 7.

@kmlebedev Thank you for your help and for adding the ability to send s3 acces logs
